diff --git a/gmp.h b/gmp-6.3.0/gmp.h
index e096026..015fab6 100644
--- a/gmp.h
+++ b/gmp-6.3.0/gmp.h
@@ -1683,6 +1683,9 @@ __GMP_DECLSPEC void mpn_sec_powm (mp_ptr, mp_srcptr, mp_size_t, mp_srcptr, mp_bi
 #define mpn_sec_powm_itch __MPN(sec_powm_itch)
 __GMP_DECLSPEC mp_size_t mpn_sec_powm_itch (mp_size_t, mp_bitcnt_t, mp_size_t) __GMP_ATTRIBUTE_PURE;
 
+#define   mpn_powm_2exp __MPN(powm_2exp)
+__GMP_DECLSPEC void      mpn_powm_2exp (mp_ptr, mp_srcptr, mp_size_t, mp_bitcnt_t, mp_srcptr, mp_size_t, mp_ptr);
+
 #define mpn_sec_tabselect __MPN(sec_tabselect)
 __GMP_DECLSPEC void mpn_sec_tabselect (volatile mp_limb_t *, volatile const mp_limb_t *, mp_size_t, mp_size_t, mp_size_t);
 
@@ -1700,6 +1703,9 @@ __GMP_DECLSPEC int mpn_sec_invert (mp_ptr, mp_ptr, mp_srcptr, mp_size_t, mp_bitc
 #define mpn_sec_invert_itch __MPN(sec_invert_itch)
 __GMP_DECLSPEC mp_size_t mpn_sec_invert_itch (mp_size_t) __GMP_ATTRIBUTE_PURE;
 
+#define mpn_binvert_itch __MPN(binvert_itch)
+__GMP_DECLSPEC mp_size_t mpn_binvert_itch (mp_size_t);
+
 
 /**************** mpz inlines ****************/
 
